File: java/src/test/java/com/epam/deltix/containers/ListFactoryTest.java
package com.epam.deltix.containers;

import org.junit.Assert;
import org.junit.Test;

import java.util.ArrayList;
import java.util.Random;

/**
 * Created by DriapkoA on 19/01/2018.
 */
public class ListFactoryTest {

    @Test
    public void stressTest() throws Exception {
        int numberOfIterations = 2000000;
        int numberOfLists = 100;
        ListFactory<Integer> factory = new ListFactory<>();
        ArrayList<Integer> ids = new ArrayList<>();
        ArrayList<BufferedLinkedList<Integer>> etalonLists = new ArrayList<>();
        for (int i = 0; i < numberOfLists; ++i) {
            ids.add(factory.createList());
            etalonLists.add(new BufferedLinkedList<Integer>());
        }
        Random random = new Random(55);
        for (int i = 0; i < numberOfIterations; ++i) {
            int type = Math.abs(random.nextInt()) % 3;
            if (type < 2) {
                int index = Math.abs(random.nextInt()) % numberOfLists;
                int listId = ids.get(index);
                int element = random.nextInt();
                int insertionType = (Math.abs(random.nextInt()) % 4);
                if (insertionType == 0) {
                    factory.addLast(listId, element);
                    etalonLists.get(index).addLast(element);
                } else if (insertionType == 1) {
                    factory.addFirst(listId, element);
                    etalonLists.get(index).addFirst(element);
                } else if (insertionType == 2) {
                    if (etalonLists.get(index).getCount() < 2) continue;
                    int secondId = factory.getNext(factory.getFirst(listId));
                    factory.addAfter(secondId, element);
                    etalonLists.get(index).addAfter(etalonLists.get(index).next(etalonLists.get(index).getFirstKey()), element);
                } else if (insertionType == 3) {
                    if (etalonLists.get(index).getCount() < 2) continue;
                    int secondId = factory.getNext(factory.getFirst(listId));
                    factory.addBefore(secondId, element);
                    etalonLists.get(index).addBefore(etalonLists.get(index).next(etalonLists.get(index).getFirstKey()), element);
                }
            } else {
                int removeType = Math.abs(random.nextInt()) % 3;
                int index = Math.abs(random.nextInt()) % numberOfLists;
                int listId = ids.get(index);
                if (removeType == 0) {
                    if (etalonLists.get(index).getCount() == 0) continue;
                    etalonLists.get(index).remove(etalonLists.get(index).getFirstKey());
                    factory.deleteElement(factory.getFirst(listId));
                } else if (removeType == 1) {
                    if (etalonLists.get(index).getCount() == 0) continue;
                    etalonLists.get(index).remove(etalonLists.get(index).getLastKey());
                    factory.deleteElement(factory.getLast(listId));
                } else if (removeType == 2) {
                    if (etalonLists.get(index).getCount() < 2) continue;
                    etalonLists.get(index).remove(etalonLists.get(index).next(etalonLists.get(index).getFirstKey()));
                    factory.deleteElement(factory.getNext(factory.getFirst(listId)));
                }
            }
            if (i % 10000 == 0) {
                for (int j = 0; j < ids.size(); ++j) {
                    int listId = ids.get(j);
                    Assert.assertEquals(etalonLists.get(j).getCount(), factory.getCount(listId));
                    int ptr1 = etalonLists.get(j).getFirstKey();
                    int ptr2 = factory.getFirst(listId);
                    int n = factory.getCount(listId);
                    for (int k = 0; k < n; ++k) {
                        Assert.assertEquals(etalonLists.get(j).getElementByKey(ptr1), factory.get(ptr2));
                        ptr1 = etalonLists.get(j).next(ptr1);
                        ptr2 = factory.getNext(ptr2);
                    }
                }
            }
        }
        for (int j = 0; j < ids.size(); ++j) {
            int listId = ids.get(j);
            Assert.assertEquals(etalonLists.get(j).getCount(), factory.getCount(listId));
            int ptr1 = etalonLists.get(j).getFirstKey();
            int ptr2 = factory.getFirst(listId);
            int n = factory.getCount(listId);
            for (int k = 0; k < n; ++k) {
                Assert.assertEquals(etalonLists.get(j).getElementByKey(ptr1), factory.get(ptr2));
                ptr1 = etalonLists.get(j).next(ptr1);
                ptr2 = factory.getNext(ptr2);
            }
        }

    }


}

File: csharp/tests/Containers.Tests/ListFactoryTest.cs
using NUnit.Framework;
using System;
using System.Collections.Generic;
using System.Text;

namespace EPAM.Deltix.Containers.Tests{
	[TestFixture()]
	class ListFactoryTest
	{
		[TestCase()]
		public void ListFactoryStressTest()
		{

			int numberOfIterations = 2000000;
			int numberOfLists = 100;
			ListFactory<int> factory = new ListFactory<int>();
			List<ListFactory<int>.List> ids = new List<ListFactory<int>.List>();
			List<BufferedLinkedList<int>> etalonLists = new List<BufferedLinkedList<int>>();
			for (int i = 0; i < numberOfLists; ++i)
			{
				ids.Add(factory.CreateList());
				etalonLists.Add(new BufferedLinkedList<int>());
			}
			Random random = new Random(55);
			for (int i = 0; i < numberOfIterations; ++i)
			{

				int type = Math.Abs(random.Next()) % 3;
				if (type < 2)
				{
					int index = Math.Abs(random.Next()) % numberOfLists;

					int element = random.Next();
					int insertionType = (Math.Abs(random.Next()) % 4);
					if (insertionType == 0)
					{
						ids[index].AddLast(element);
						etalonLists[index].AddLast(element);
					}
					else if (insertionType == 1)
					{
						ids[index].AddFirst(element);
						etalonLists[index].AddFirst(element);
					}
					else if (insertionType == 2)
					{
						if (etalonLists[index].Count < 2) continue;
						ids[index].First.Next.Value.AddAfter(element);
						etalonLists[index].AddAfter(etalonLists[index].Next(etalonLists[index].FirstKey), element);
					}
					else if (insertionType == 3)
					{
						if (etalonLists[index].Count < 2) continue;
						ids[index].First.Next.Value.AddBefore(element);
						etalonLists[index].AddBefore(etalonLists[index].Next(etalonLists[index].FirstKey), element);
					}
				}
				else
				{
					int removeType = Math.Abs(random.Next()) % 3;
					int index = Math.Abs(random.Next()) % numberOfLists;
					
					if (removeType == 0)
					{
						if (etalonLists[index].Count == 0) continue;
						etalonLists[index].Remove(etalonLists[index].FirstKey);
						ids[index].First.Remove();
					}
					else if (removeType == 1)
					{
						if (etalonLists[index].Count == 0) continue;
						etalonLists[index].Remove(etalonLists[index].LastKey);
						ids[index].Last.Remove();
					}
					else if (removeType == 2)
					{
						if (etalonLists[index].Count < 2) continue;
						etalonLists[index].Remove(etalonLists[index].Next(etalonLists[index].FirstKey));
						ids[index].First.Next.Value.Remove();
					}
				}
				if (i % 10000 == 0)
				{
					
					for (int j = 0; j < ids.Count; ++j)
					{
						Assert.AreEqual(etalonLists[j].Count, ids[j].Count);
						int ptr1 = etalonLists[j].FirstKey;
						var ptr2 = ids[j].First;
						int n = ids[j].Count;
						for (int k = 0; k < n; ++k)
						{
							Assert.AreEqual(etalonLists[j][ptr1], ptr2.Data);
							ptr1 = etalonLists[j].Next(ptr1);
							if (k != n - 1) ptr2 = ptr2.Next.Value;
						}
					}
				}
			}
			for (int j = 0; j < ids.Count; ++j)
			{
				Assert.AreEqual(etalonLists[j].Count, ids[j].Count);
				int ptr1 = etalonLists[j].FirstKey;
				var ptr2 = ids[j].First;
				int n = ids[j].Count;
				for (int k = 0; k < n; ++k)
				{
					Assert.AreEqual(etalonLists[j][ptr1], ptr2.Data);
					ptr1 = etalonLists[j].Next(ptr1);
					if (k != n - 1) ptr2 = ptr2.Next.Value;
				}
			}
		}
	}
}
